packages/react-reconciler/src/ReactFiberNewContext.js

```diff
@@ -90,17 +97,178 @@ function popProvider(providerFiber: Fiber): void {
+                alternate.expirationTime > renderExpirationTime)
+            ) {
+              alternate.expirationTime = renderExpirationTime;
+            } else {
+              // Neither alternate was updated, which means the rest of the
+              // ancestor path already has sufficient priority.
+              break;
+            }
+            node = node.return;
+          }
+          // Don't scan deeper than a matching consumer. When we render the
+          // consumer, we'll continue scanning from that point. This way the
+          // scanning work is time-sliced.
+          nextFiber = null;
+        } else {
+          nextFiber = fiber.child;
+        }
+        reader = reader.next;
+      } while (reader !== null);
+    } else if (fiber.tag === ContextProvider) {
+      // Don't scan deeper if this is a matching provider
+      nextFiber = fiber.type === workInProgress.type ? null : fiber.child;
+    } else {
+      // Traverse down.
+      nextFiber = fiber.child;
+    }
+
+    if (nextFiber !== null) {
+      // Set the return pointer of the child to the work-in-progress fiber.
+      nextFiber.return = fiber;
+    } else {
+      // No child. Traverse to next sibling.
+      nextFiber = fiber;
+      while (nextFiber !== null) {
+        if (nextFiber === workInProgress) {
+          // We're back to the root of this subtree. Exit.
+          nextFiber = null;
+          break;
+        }
+        let sibling = nextFiber.sibling;
+        if (sibling !== null) {
+          // Set the return pointer of the sibling to the work-in-progress fiber.
+          sibling.return = nextFiber.return;
+          nextFiber = sibling;
+          break;
+        }
+        // No more siblings. Traverse up.
+        nextFiber = nextFiber.return;
+      }
+    }
+    fiber = nextFiber;
+  }
+}
+
+export function checkForPendingContext(
+  workInProgress: Fiber,
+  renderExpirationTime: ExpirationTime,
+): boolean {
+  let reader = workInProgress.firstContextReader;
+  let hasPendingContext = false;
+  while (reader !== null) {
+    const context = reader.context;
+    const changedBits = isPrimaryRenderer
+      ? context._changedBits
+      : context._changedBits2;
+    if (changedBits !== 0) {
+      // Resume context change propagation. We need to call this even if
+      // this fiber bails out, in case deeply nested consumers observe more
+      // bits than this one.
+      propagateContextChange(
+        workInProgress,
+        context,
+        changedBits,
+        renderExpirationTime,
+      );
+      if ((changedBits & reader.observedBits) !== 0) {
+        hasPendingContext = true;
+      }
+    }
+    reader = reader.next;
+  }
+  return hasPendingContext;
 }
 
-function getContextChangedBits(context: ReactContext<any>): number {
-  return isPrimaryRenderer ? context._changedBits : context._changedBits2;
+export function prepareToReadContext(): void {
+  nextFirstReader = nextLastReader = null;
```

Comment:
You do this in finishReadingContext also. Only it do it once per loop iteration. Either before the loop and in finishReadingContext or in prepareToReadContext. If you keep prepareToReadContext, we should inline all callsites. I don't trust tooling to make this fast.